waiter/src/waiter/util/moving_average.clj

```diff
@@ -0,0 +1,66 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns waiter.util.moving-average)
+
+(defprotocol MovingAverage
+  "Data structure that computes moving averages."
+  (insert-value [_ value] "Update the moving average with `value` and return a new MovingAverage instance.")
+  (get-average [_] "Returns the moving average represented by this instance."))
+
+;; Computes the simple moving average over a defined window size.
+(defrecord SlidingWindowMovingAverage [window-size data data-sum]
```

Comment:
Did not use [`com.codahale.metrics.SlidingWindowReservoir`](https://github.com/dropwizard/metrics/blob/3680c67b2b0db0346a2273fcbcaa80a7d29aacd8/metrics-core/src/main/java/com/codahale/metrics/SlidingWindowReservoir.java#L9) as computing the average requires [copying and sorting the data array](https://github.com/dropwizard/metrics/blob/3680c67b2b0db0346a2273fcbcaa80a7d29aacd8/metrics-core/src/main/java/com/codahale/metrics/UniformSnapshot.java#L24).